arangod/Cluster/ParticipantsCacheFeature.cpp

```diff
@@ -0,0 +1,251 @@
+  auto agencyCallbackRegistry =
+      server().getEnabledFeature<ClusterFeature>().agencyCallbackRegistry();
+  if (agencyCallbackRegistry == nullptr) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "Expected non-null AgencyCallbackRegistry "
+                                   "when starting ParticipantsCacheFeature.");
+  }
+
+  initHealthCache();
+  _cache->start(agencyCallbackRegistry);
+
+  scheduleFlush();
+  LOG_TOPIC("42af3", DEBUG, Logger::REPLICATION2)
+      << "ParticipantsCacheFeature is ready";
+}
+
+void ParticipantsCacheFeature::stop() {
+  LOG_DEVEL << "ParticipantsCacheFeature stopped";
+  try {
+    auto agencyCallbackRegistry =
+        server().getEnabledFeature<ClusterFeature>().agencyCallbackRegistry();
+    _cache->stop(agencyCallbackRegistry);
+  } catch (std::exception const& ex) {
+    LOG_TOPIC("42af2", WARN, Logger::REPLICATION2)
+        << "caught unexpected exception while unregistering agency callback in "
+           "ParticipantsCacheFeature: "
+        << ex.what();
+  }
+  if (_flushJob) {
+    _flushJob->cancel();
+  }
+}
+
+auto ParticipantsCacheFeature::status()
+    -> std::unordered_map<std::string, bool> {
+  return _cache->getIsFailed();
+}
+
+void ParticipantsCacheFeature::flush() {
+  LOG_DEVEL << "ParticipantsCacheFeature flushed";
+  AgencyCache& agencyCache =
+      server().getEnabledFeature<ClusterFeature>().agencyCache();
+  std::shared_ptr<VPackBuilder> builder;
+  std::tie(builder, std::ignore) = agencyCache.get(kSupervisionHealthPath);
+  if (auto result = builder->slice(); !result.isNone()) {
+    TRI_ASSERT(result.isObject())
+        << " expected object in agency at " << kSupervisionHealthPath
+        << " but got " << result.toString();
+    FailureMap newMap;
+    for (auto const& [key, value] : VPackObjectIterator(result)) {
+      auto serverId = key.copyString();
+      auto isServerGood =
+          value.get(kHealthyServerKey).isEqualString(HealthyServerValue);
+      newMap[serverId] = !isServerGood;
+    }
+    _cache->reset(std::move(newMap));
+  }
+}
+
+auto ParticipantsCacheFeature::getFailureOracle()
+    -> std::shared_ptr<IFailureOracle> {
+  return std::static_pointer_cast<IFailureOracle>(_cache);
+}
+
+void ParticipantsCacheFeature::initHealthCache() {
+  TRI_ASSERT(_cache == nullptr);
+
+  _cache = std::make_shared<ParticipantsCache>();
+  _cache->createAgencyCallback(server());
+}
+
+void ParticipantsCacheFeature::scheduleFlush() {
+  using namespace std::chrono_literals;
+
+  auto scheduler = SchedulerFeature::SCHEDULER;
+  if (!scheduler) {
+    return;
+  }
+
+  _flushJob = scheduler->queueDelayed(RequestLane::AGENCY_CLUSTER, 50s,
+                                      [this](bool canceled) {
+                                        if (!canceled) {
+                                          this->flush();
+                                          this->scheduleFlush();
+                                        }
```

Comment:
Is this code actually ok? I think there can be a race between calling `cancel` and this lambda being executed. The feature might also be deleted (I am not sure about that part). But I am pretty sure that the lambda can still be executed after `stop` returned. (which `cancel == false`)